contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,486 @@
+Vertical Pod Autoscaler
+=======================
+Vertical Pod Autoscaler (VPA, aka. "rightsizing" or "autopilot") is an
+infrastructure service that automatically sets resource requirements of Pods and
+dynamically adjusts them in runtime, based on analysis of historical resource
+utilization, amount of resources available in the cluster and real-time events,
+such as OOMs.
+
+- [Introduction](#introduction)
+  - [Background](#background)
+  - [Purpose](#purpose)
+  - [Related features](#related-features)
+- [Requirements](#requirements)
+  - [Functional](#functional)
+  - [Availability](#availability)
+  - [Extensibility](#extensibility)
+- [Design](#design)
+  - [Overview](#overview)
+  - [Architecture overview](#architecture-overview)
+  - [API](#api)
+  - [Admission Controller](#admission-controller)
+  - [Life of the Pod](#life-of-the-pod)
+  - [Recommender](#recommender)
+  - [Updater](#updater)
+  - [Recommendation model](#recommendation-model)
+  - [History Storage](#history-storage)
+  - [Open questions](#open-questions)
+- [Alternatives considered](#alternatives-considered)
+  - [Pods point at VPA](#pods-point-at-vpa)
+  - [VPA points at Deployment](#vpa-points-at-deployment)
+  - [Actuation using the Deployment update mechanism](#actuation-using-the-deployment-update-mechanism)
+  
+------------
+Introduction
+------------
+
+### Background ###
+* [Compute resources](http://kubernetes.io/docs/user-guide/compute-resources/)
+* [Resource QoS](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-qos.md)
+* [Admission Controllers](http://kubernetes.io/docs/admin/admission-controllers/)
+
+### Purpose ###
+Vertical scaling has two objectives: improving utilization of compute resources
+and reducing the maintenance cost, by automating configuration of resource
+requirements. While ensuring high utilization it tries to minimize the risk of
+containers running out of memory or getting CPU starved.
+
+### Related features ###
+#### Initial resources ####
+["Initial Resources"](https://github.com/kgrygiel/community/blob/master/contributors/design-proposals/initial-resources.md)
+is a very preliminary, proof-of-concept feature providing initial request based
+on historical utilization. It is designed to only kick in on Pod creation.
+VPA is intended to supersede this feature.
+
+#### In-place updates ####
+In-place Pod updates is a planned feature to allow changing resources
+(request/limit) of existing containers without killing them, assuming sufficient
+free resources available on the node. Vertical Pod Autoscaler will greatly
+benefit from this ability, however it is not considered a blocker for the MVP.
+#### Resource estimation ####
```

Comment:
@kgrygiel thoughts on this comment? Given that kubelet can make vertical scaling decisions prior to a container getting OOM killed or running out of storage with low latency, it is beneficial to have the multi-tier model that @bgrant0607 suggested.